### PR TITLE
Add Read Update Delete tests for ThreadPools and an end-to-end integration test

### PR DESCRIPTION
**Pull request by lloydmeta**
- For admin ThreadPools, During the 2.4.x migration, the Create integration test was already added.
- This PR also adds an end-to-end integration test suite (listing Parts, asking for parts)

### CODE REVIEW

File: project/Octoparts.scala [hunk 1]
```diff
@@ -20,7 +20,7 @@ object Octoparts extends Build {
   // -------------------------------------------------------
   lazy val app = Project(id = "octoparts", base = file("."), settings = playAppSettings)
     .enablePlugins(PlayScala)
-    .dependsOn(models, authHandlerApi, playJsonFormats)
+    .dependsOn(models, authHandlerApi, playJsonFormats, scalaWsClient % "test->compile")
```
Comment (xevix) [inline project/Octoparts.scala:23]: Could you explain this particular change?

File: test/integration/AdminSpec.scala [hunk 2]
```diff
@@ -18,15 +18,100 @@ class AdminSpec
...
+        descriptors.find(_.text == size.toString) shouldBe 'defined
+        descriptors.find(_.text == queueSize.toString) shouldBe 'defined
+      }
+
+    }
+
+    describe("the show Thread Pool page") {
+
+      it("should redirect me to the thread poollist page if no such thread pool can be found") {
+        goTo(ThreadPoolShowPage(35415151413L))
```
Comment (xevix) [inline test/integration/AdminSpec.scala]: Is this just an arbitrarily-big Long?

File: project/Octoparts.scala (hunk 1 above)
Comment (lloydmeta) [inline project/Octoparts.scala:23]: Yeah, this is a cryptic addition that adds a test-mode dependency in the current project definition (root app in this case); in particular the ScalaWsClient project is to be in a compiled state before running tests in root.

The reason I did this was to allow us to integration test the root project (Octoparts app) using the ScalaWsClient.

File: test/integration/AdminSpec.scala (hunk 2 above)
Comment (lloydmeta) [inline test/integration/AdminSpec.scala]: Yes.

Comment (lloydmeta) [inline test/integration/AdminSpec.scala]: After discussion, will use `Long.MaxValue`

Comment (lloydmeta) [inline test/integration/AdminSpec.scala]: Done.

File: project/Octoparts.scala (hunk 1 above)
Comment (xevix) [inline project/Octoparts.scala:23]: Ah I see. Maybe a one-liner comment to explain this would be good then.

Comment (lloydmeta) [inline project/Octoparts.scala:23]: Good idea. Done :+1:

File: test/integration/AdminSpec.scala [hunk 3]
```diff
@@ -36,7 +121,7 @@ class AdminSpec
     describe("adding a Part") {
 
       it("should work and redirect me to the part show page") {
-        val PartAddPage.PartConfig(partId, url, connectionPoolSize, commandKey, commandKeyGroup, proxy) = PartAddPage.createPart(ThreadPoolAddPage.createThreadPool)
+        val PartAddPage.PartConfig(partId, url, connectionPoolSize, commandKey, commandKeyGroup, Some(proxy)) = PartAddPage.createPart(ThreadPoolAddPage.createThreadPool._2)
```
Comment (xevix) [inline test/integration/AdminSpec.scala]: Perhaps split the tuple into its own line for clarity, as `._2` is not immediately obvious.

``` scala
val (_, threadPoolConfig) = ThreadPoolAddPage.createThreadPool
val PartAddPage.PartConfig(partId, _, _, _, _, _) = PartAddPage.createPart(threadPoolConfig)
```

Likewise for L143.

Comment (lloydmeta) [inline test/integration/AdminSpec.scala]: Good idea. Done.

File: app/com/m3/octoparts/util/OctoMetricsImpl.scala [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+/**
+ * During tests, we occasionally have cases where the asynchronous shutting down of an application means that
+ * the previous registry in the singleton has not been unregistered before starting the next test.
+ *
+ * This implementation of Metrics overrides onStart to perform onStop cleanups first before proceeding with
+ * other onStart duties
+ */
+class OctoMetricsImpl(lifecycle: ApplicationLifecycle, configuration: Configuration) extends MetricsImpl(lifecycle, configuration) {
+
+  override def onStart(): ObjectMapper = {
```
Comment (xevix) [inline app/com/m3/octoparts/util/OctoMetricsImpl.scala:19]: Is it possible to add a test for this, or is it too edge-casey? I suppose you'd just be testing to see if the metric was removed from the registry.

Although the comment above mentions that this is to address issues when running tests, this implementation with the override is being used in non-test code (`UtilsModule`) in place of the default `MetricsImpl`. In this case, I think that either this needs testing, or that the custom implementation should only be used in testing.

Comment (lloydmeta) [inline app/com/m3/octoparts/util/OctoMetricsImpl.scala:19]: I think you're right in saying that this custom implementation should be tested in order to make sure it does what it is supposed to do: remove metrics with the same names from the registry when `onStart` is run and _then_ go about adding metrics into the registry. I think it should be enough to add a test that does the following:
1. Sets up the global registry singleton to have the said metrics
2. Instantiates OctoMetricsImpl
3. Calls onStart
4. Ensures the metrics are in the registry
   - Implicitly, we know the custom implementation works because no error has been thrown during 3.

What do you think?

Comment (xevix) [inline app/com/m3/octoparts/util/OctoMetricsImpl.scala:19]: Calling `onStart` attempts to remove the metric from the internal map. I'm guessing the error thrown would be from the concurrent map [in use here in Codahale Metrics](https://github.com/dropwizard/metrics/blob/d8d8e2796d1f7dfb5268e2e8fbf35039570c53d7/metrics-core/src/main/java/io/dropwizard/metrics/SharedMetricRegistries.java#L26) itself [called from Kenshoo Metrics](https://github.com/kenshoo/metrics-play/blob/master/src/main/scala/com/kenshoo/play/metrics/Metrics.scala#L80). What would it throw specifically, that it could not remove a key not in the map?

Comment (lloydmeta) [inline app/com/m3/octoparts/util/OctoMetricsImpl.scala:19]: `onStart` with this custom implementation (`OctoMetricsImpl`) will try to remove the metrics from the internal metrics registry map by invoking the parent `onStop` cleanup method before proceeding with normal `onStart` duties; therefore, should be no error thrown; still, a good idea to catch NonFatal errors.

The default implementation of `onStart` in `MetricsImpl` will not try to clean up first; it will therefore throw an error if there are already metrics registered in the global singleton with the same names as the ones that it tries to register (`IllegalArgumentException`)

Comment (xevix) [inline app/com/m3/octoparts/util/OctoMetricsImpl.scala:19]: Ah ok, so the catch here was more for paranoia (which I'm totally for when it comes to internal config issues).

Yeah, this plan sounds sound (ugh).

Comment (lloydmeta) [inline app/com/m3/octoparts/util/OctoMetricsImpl.scala:19]: Done in af1f508

File: test/com/m3/octoparts/util/OctoMetricsImplSpec.scala [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+  lazy val subject = new OctoMetricsImpl(appComponents.applicationLifecycle, appComponents.configuration)
+
+  describe("#onStart") {
+
+    it("should not throw") {
+      subject.onStart()
+    }
+
+    it("should work with consecutive calls") {
```
Comment (xevix) [inline test/com/m3/octoparts/util/OctoMetricsImplSpec.scala]: I'm bordering on pedantry, but does "should work" mean "should not throw" in this context?

File: test/com/m3/octoparts/util/OctoMetricsImplSpec.scala [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+      subject.onStart()
+      subject.onStart()
+    }
+
+    it("should work if metrics were already registered") {
```
Comment (xevix) [inline test/com/m3/octoparts/util/OctoMetricsImplSpec.scala]: More pedantry and my own confirmation, this is checking if the _same_ metrics were already registered on the same default registry, yes?

A slight suggestion would be to create the subject with `jvmMetricsEnabled` true, and just call `setupJvmMetrics` before the `onStart` call. Unfortunately the specific attributes that get enabled aren't factored into a list in Kenshoo so the `find`s will have to remain =/

File: test/com/m3/octoparts/util/OctoMetricsImplSpec.scala (hunk 5 above)
Comment (lloydmeta) [inline test/com/m3/octoparts/util/OctoMetricsImplSpec.scala]: Yes. Renamed for clarity.

File: test/com/m3/octoparts/util/OctoMetricsImplSpec.scala (hunk 6 above)
Comment (lloydmeta) [inline test/com/m3/octoparts/util/OctoMetricsImplSpec.scala]: The original intent was to make it obvious within the test that _these_ are the metrics we want to ensure are registered before calling `.onStart`, and to make that expectation explicit.

However, thinking about it again, it's just as easy to make it explicit by refactoring out the check into a separate method, call that method after calling `setupJvmMetrics`, and then call it again after calling `.onStart`. Done as part of 7a1e4fc

Comment (xevix) [inline test/com/m3/octoparts/util/OctoMetricsImplSpec.scala]: Great! I hope this didn't sound too pedantic, thanks for the refactor.

**Comment by xevix** [comment]
LGTM. Does this merit a version bump and a release?

**Comment by lloydmeta** [comment]
Nah, I don't think so; [this line that modifies the header check](https://github.com/m3dev/octoparts/pull/192/files#diff-92f67571d02b81ad83359da68946b9eeR67) is the only impactful change in running-code and I doubt we'll see noticeable performance improvements from that :laughing:
